### ISSUE TITLE
LogicError: cuCtxSynchronize failed: an illegal memory access was encountered

### ISSUE THREAD

**Issue by ricpruss** [cair/PyTsetlinMachineCUDA#3]
```
tm = MultiClassTsetlinMachine(number_of_clauses=8000, T=100*100, s=10.0, max_weight=1)
print("\nAccuracy over 100 epochs:\n")
for i in tqdm(range(100)):
	start_training = time()
	tm.fit(X_train, Y_train, epochs=1, incremental=True)
	stop_training = time()

	start_testing = time()
	result = 100*(tm.predict(X_test) == Y_test).mean()
	stop_testing = time()

	print("#%d Accuracy: %.2f%% Training: %.2fs Testing: %.2fs" % (i+1, result, stop_training-start_training, stop_testing-start_testing))

```
Some versions...
```

-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.100      Driver Version: 440.100      CUDA Version: 10.2     |

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Wed_Oct_23_19:24:38_PDT_2019
Cuda compilation tools, release 10.2, V10.2.89

gcc version 8.4.0 (Ubuntu 8.4.0-3ubuntu2) 
```
The output around the crash:
```

Accuracy over 100 epochs:

HBox(children=(FloatProgress(value=0.0), HTML(value='')))

#1 Accuracy: 37.04% Training: 105.69s Testing: 14.27s
#2 Accuracy: 46.91% Training: 122.00s Testing: 15.30s
#3 Accuracy: 50.12% Training: 123.95s Testing: 15.78s
#4 Accuracy: 49.88% Training: 123.49s Testing: 16.26s
#5 Accuracy: 50.37% Training: 123.60s Testing: 16.00s
#6 Accuracy: 51.36% Training: 123.97s Testing: 15.83s
#7 Accuracy: 49.88% Training: 123.88s Testing: 15.59s
#8 Accuracy: 50.37% Training: 123.09s Testing: 15.39s
#9 Accuracy: 50.86% Training: 122.19s Testing: 15.16s
#10 Accuracy: 49.63% Training: 121.56s Testing: 14.92s
#11 Accuracy: 51.60% Training: 121.21s Testing: 14.63s

---------------------------------------------------------------------------
LogicError                                Traceback (most recent call last)
<ipython-input-26-f4fe9c9c223b> in <module>
      2 for i in tqdm(range(100)):
      3         start_training = time()
----> 4         tm.fit(X_train, Y_train, epochs=1, incremental=True)
      5         stop_training = time()
      6 

~/github/PyTsetlinMachineCUDA/PyTsetlinMachineCUDA/tm.py in fit(self, X, Y, epochs, incremental, batch_size)
    313                         self.ctm = MultiClassConvolutionalTsetlinMachine2D(self.number_of_clauses, self.T, self.s, self.patch_dim, boost_true_positive_feedback=self.boost_true_positive_feedback, number_of_state_bits=self.number_of_state_bits, append_negated=self.append_negated, max_weight=self.max_weight, number_of_classes=self.number_of_classes)
    314 
--> 315                 self.ctm.fit(X, Y, epochs = epochs, incremental = incremental, batch_size = batch_size)
    316 
    317                 return

~/github/PyTsetlinMachineCUDA/PyTsetlinMachineCUDA/tm.py in fit(self, X, Y, epochs, incremental, batch_size)
    235                         for e in range(0, number_of_examples, batch_size):
    236                                 self.update.prepared_call((16*13,1,1), (128,1,1), g.state, self.ta_state_gpu, self.clause_weights_gpu, self.class_sum_gpu, self.clause_output_gpu, self.clause_patch_gpu, self.encoded_X_training_gpu, self.Y_gpu, np.int32(e))
--> 237                                 cuda.Context.synchronize()
    238 
    239                 self.ta_state = np.array([])

LogicError: cuCtxSynchronize failed: an illegal memory access was encountered
```

**Comment by ricpruss** [comment]
Intermittent.

**Comment by olegranmo** [comment]
Thanks for reporting, @ricpruss! If possible, it would be great if you can share your code and I will try to reconstruct on my side.

**Comment by ricpruss** [comment]
I used your example for MultiClassTsetlinMachine verbatim but with different data and then changed the code for the clauses to what I cut and pasted about.

**Comment by ricpruss** [comment]
I just noticed that T=100*100 would be greater than the number of clauses... would that cause the issue?

**Comment by olegranmo** [comment]
Thanks for the feedback. Large T should not cause this problem. One explanation could be that memory is accessed outside the allocated area. Is the shape of X and Y still the same as for the MNIST data?

**Comment by desertnaut** [comment]
Had a similar issue (see link above), which was resolved with the new PyTsetlinMachineCUDA version **0.1.8**.

**Comment by ricpruss** [comment]
Confirmed this seems closed in 0.1.8.
